src/Workleap.Extensions.Mongo/Telemetry/DiagnosticsTracingEventSubscriber.cs

```diff
@@ -0,0 +1,152 @@
+
+    private void Handle(ConnectionSendingMessagesFailedEvent @event) => this.HandleDiagnosticEvent(@event);
+
+    private void Handle(ConnectionSentMessagesEvent @event) => this.HandleDiagnosticEvent(@event);
+    #endregion
+
+    #region connection pool events
+    private void Handle(ConnectionPoolAddedConnectionEvent @event) => this.HandleDiagnosticEvent(@event);
+
+    private void Handle(ConnectionPoolAddingConnectionEvent @event) => this.HandleDiagnosticEvent(@event);
+
+    private void Handle(ConnectionPoolCheckedInConnectionEvent @event) => this.HandleDiagnosticEvent(@event);
+
+    private void Handle(ConnectionPoolCheckedOutConnectionEvent @event) => this.HandleDiagnosticEvent(@event);
+
+    private void Handle(ConnectionPoolCheckingInConnectionEvent @event) => this.HandleDiagnosticEvent(@event);
+
+    private void Handle(ConnectionPoolCheckingOutConnectionEvent @event) => this.HandleDiagnosticEvent(@event);
+
+    private void Handle(ConnectionPoolCheckingOutConnectionFailedEvent @event) => this.HandleDiagnosticEvent(@event);
+
+    private void Handle(ConnectionPoolClearedEvent @event) => this.HandleDiagnosticEvent(@event);
+
+    private void Handle(ConnectionPoolClearingEvent @event) => this.HandleDiagnosticEvent(@event);
+
+    private void Handle(ConnectionPoolClosedEvent @event) => this.HandleDiagnosticEvent(@event);
+
+    private void Handle(ConnectionPoolClosingEvent @event) => this.HandleDiagnosticEvent(@event);
+
+    private void Handle(ConnectionPoolOpenedEvent @event) => this.HandleDiagnosticEvent(@event);
+
+    private void Handle(ConnectionPoolOpeningEvent @event) => this.HandleDiagnosticEvent(@event);
+
+    private void Handle(ConnectionPoolReadyEvent @event) => this.HandleDiagnosticEvent(@event);
+
+    private void Handle(ConnectionPoolRemovedConnectionEvent @event) => this.HandleDiagnosticEvent(@event);
+
+    private void Handle(ConnectionPoolRemovingConnectionEvent @event) => this.HandleDiagnosticEvent(@event);
+    #endregion
+
+    #region server events
+    private void Handle(ServerClosedEvent @event) => this.HandleDiagnosticEvent(@event);
+
+    private void Handle(ServerClosingEvent @event) => this.HandleDiagnosticEvent(@event);
+
+    private void Handle(ServerDescriptionChangedEvent @event) => this.HandleDiagnosticEvent(@event);
+
+    private void Handle(ServerHeartbeatFailedEvent @event) => this.HandleDiagnosticEvent(@event);
+
+    private void Handle(ServerHeartbeatStartedEvent @event) => this.HandleDiagnosticEvent(@event);
+
+    private void Handle(ServerHeartbeatSucceededEvent @event) => this.HandleDiagnosticEvent(@event);
+
+    private void Handle(ServerOpenedEvent @event) => this.HandleDiagnosticEvent(@event);
+
+    private void Handle(ServerOpeningEvent @event) => this.HandleDiagnosticEvent(@event);
+    #endregion
+
+    private void HandleDiagnosticEvent(object diagnosticEvent)
+    {
+        if (!this._options.Telemetry.CaptureDiagnosticEvents || Activity.Current is null)
```

Comment:
Can we capture the activity here and use it later when adding the tags.

It's because we do a null check here, then some computing and again a null check when adding the tags.